src/Router.php

```diff
@@ -119,44 +119,42 @@ protected function matchDirect(string $httpMethod, string $urlToMatch): ?Route
         return $route;
     }
 
-    protected function matchParametered(string $httpMethod, string $urlToMatch): ?Route
+    protected function matchParameteredOrCatchAll(string $httpMethod, string $urlToMatch): ?Route
     {
         if (!array_key_exists($httpMethod, $this->routes)) {
             return null;
         }
 
         foreach ($this->routes[$httpMethod] as $routeUrl => $route) {
-            if (!str_contains($routeUrl, '{') || !$route->supportsHttpMethod($httpMethod)) {
-                continue;
-            }
-
             $explodedUrlToMatch = explode('/', trim($urlToMatch, '/'));
             $explodedRouteUrl = explode('/', trim($routeUrl, '/'));
 
-            if (count($explodedUrlToMatch) !== count($explodedRouteUrl)) {
-                continue;
-            }
-
-            $providedValues = array_diff($explodedUrlToMatch, $explodedRouteUrl);
-
             $valuesWithParameterAsKey = [];
+            $catchAllValues = [];
+            $catchAllParameterIndex = 0;
+            foreach ($explodedUrlToMatch as $key => $value) {
+                $parameter = $explodedRouteUrl[$key] ?? null;
 
-            foreach ($providedValues as $key => $value) {
-                $parameter = $explodedRouteUrl[$key];
-
-                if (!str_contains($parameter, '{')) {
+                if ($parameter === null && !$route->hasCatchAll()) {
                     continue;
                 }
 
-                $valuesWithParameterAsKey[trim($parameter, '{}')] = $value;
-
-                $explodedUrlToMatch[$key] = $parameter;
+                if (!$parameter || $parameter === '*') {
+                    $catchAllValues[$catchAllParameterIndex++] = $value;
```

Comment:
How is the `++` with a counter logic necessary here? Wouldn't `$c[] = $value` do the same?